feat(cli): Support GitHub Workflow Commands

nikeee opened the pull request:
## Summary
See discussion in #441

## WIP
Open questions / TODO:

- [x] Where to put the env var check? Maybe add a field to `TraversalMode::CI` that states if the cli is running in some CI (determined by the env vars).
- [x] File name escapes

ematipico commented:
> Where to put the env var check? Maybe add a field to `TraversalMode::CI` that states if the cli is running in some CI (determined by the env vars).

I think so. We could add something like

```rs
    CI {
        /// Whether the CI is running in a specific environment, e.g. GitHub, GitLab, etc.
        environment_name: Option<String>
    },
```
We could create a new function to read the environment variable:

```rs

    pub(crate) fn new_ci() -> Self {
        let environment_name = std::option_env!("GITHUB_ACTIONS").and_then(|_| {
            Some(String::from("GITHUB_ACTIONS"))
        });
        Self {
            report_mode: ReportMode::default(),
            traversal_mode: TraversalMode::CI {
                environment_name
            },
            max_diagnostics: MAXIMUM_DISPLAYABLE_DIAGNOSTICS,
        }
    }
```
This will be called in the `ci.rs` file


> 
> * File name escapes

ematipico commented on crates/biome_cli/src/execute/traverse.rs:
```diff
@@ -567,12 +568,19 @@ fn process_messages(options: ProcessMessagesOptions) {
         });
     }
 
+    let emit_gh_diags = mode.is_ci() && mode.should_report_to_terminal() && mode.is_running_in_ci();
+    if emit_gh_diags {
+        for diagnostic in diagnostics_to_print {
+            console.log(markup! {{PrintGitHubDiagnostic::simple(&diagnostic)}});
+        }
+    }
+
```
The `diagnostics_to_print` diagnostics are already filtered with the `should_report_to_terminal`, so it's unnecessary here.

Also, there's no need for a second loop, we can already use the loop we have at line 571

nikeee replied:
I did a second loop, so the workflow commands are printed after all the human-readable results.
Of course, we can just emit the commands after each diagnostic.
Did that in 49651245ea.

ematipico replied:
Oh, I see. Do you think it makes a difference? 🤔 I suppose we'll have to experiment and see?

nikeee replied:
Both get printed to different outputs (stderr on diagnositcs vs stdout for workflow commands). When printed on a terminal, this might cause broken error messages because the terminal renders both at the same time (and their buffers bay be flushed at different times). But in CI, it shouldn't make any difference.

nikeee commented:
Thanks for the feedback. I've updated the PR.

However, I save the environment as an enum:
```rs
pub(crate) enum ExecutionEnvironment {
    GitHub,
}
```

I also used `std::env::var` instead of `std::option_env!` because the latter is a compile-time check and we need a runtime check here.

> I'm not sure what you mean; can you explain?

As the workflow commands are just "CSV" [^1], we need to escape `,` in file names, so the fields/files are not mixed up. At least that's what I thought.
I've looked in the [documentation](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions) as well as the [`core library source`](https://github.com/actions/toolkit/blob/fe3e7ce9a7f995d29d1fcfd226a32bca407f9dc8/packages/core/src/utils.ts#L26-L41) and did not find any sign of escaping or quoting. Maybe we can just omit this and assume that the file name does not contain a `,`.


I still haven't managed to get the diagnostic description due to the printer function using a biome_console `Formatter` and the `diagnostic.description(fmt)` expecting an std formatter.


We also need to set the env var `GITHUB_ACTIONS` in CI, so this feature doesn't break the current tests (since it emits more than before ;)). Setting it to `false` should work. Or we just update the snapshots.


[1]: `"::{} file={},line={},endLine={},col={},endColumn={}::{}"`

ematipico commented on crates/biome_diagnostics/src/display_github.rs:
```diff
@@ -0,0 +1,107 @@
+        let command = match diagnostic.severity() {
+            Severity::Error | Severity::Fatal => "error",
+            Severity::Warning => "warning",
+            Severity::Hint => "notice",
+            Severity::Information => return Ok(()),
```
For now, `Hint` and `Information` should fallback to `"notice"`

ematipico commented on crates/biome_diagnostics/src/display_github.rs:
```diff
@@ -0,0 +1,107 @@
+            Severity::Hint => "notice",
+            Severity::Information => return Ok(()),
+        };
+
```
To print the advices, you can create a formatter right here, like in this example:

https://github.com/biomejs/biome/blob/b0cdf485f885b3a6ac4eda17bc956c113005b358/crates/biome_diagnostics/src/display.rs#L271-L272

Then, create a new struct similar to this one. This new struct will be responsible to write the message. Usually, the message:

https://github.com/biomejs/biome/blob/b0cdf485f885b3a6ac4eda17bc956c113005b358/crates/biome_diagnostics/src/display.rs#L331

Here's where we write the first message of a diagnostic. In this case we don't need to do anything special, but we should follow the same logic:

https://github.com/biomejs/biome/blob/b0cdf485f885b3a6ac4eda17bc956c113005b358/crates/biome_diagnostics/src/display.rs#L264-L293

nikeee replied:
Do you mean using somethign like this as a buffer?
```rs
let title = {
    let mut message = String::new();
    let mut fmt = fmt::Formatter::new(&mut message);
    fmt.write_markup(markup!({ PrintDiagnosticMessage(diagnostic) }))?;
    message
};

// ...
struct PrintDiagnosticMessage<'fmt, D: ?Sized>(&'fmt D);

impl<'fmt, D: Diagnostic + ?Sized> fmt::Display for PrintDiagnosticMessage<'fmt, D> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> io::Result<()> {
        let Self(diagnostic) = *self;
        diagnostic.message(fmt)?;
        Ok(())
    }
}
```

There doesn't seem to be a `biome_console::fmt::Write` implementation for `String`, so this doesn't seem to work. Or do I have to get the String from MarkupBuf somehow? Or did I get something completely wrong?

ematipico replied:
Actually, we have a utility that we usually use in our tests/scripts:

https://github.com/biomejs/biome/blob/e3fc59d8ceed28c8df4e5d19cf80c12cda120142/xtask/lintdoc/src/main.rs#L888-L896

We can copy it and adapt it. For example, we should gracefully handle the `expect`

nikeee replied:
I tried doing that in 2d9df8a11f7e58f3a4b02fe07e151636cf09cc7a, can you take a look?

ematipico replied:
It looks good to me as the first implementation. I am sure there's a better way, but I need to my head into it to find a safer way :)

nikeee commented:
@ematipico the tests are failing due to the workflow commands being emitted because the whole test suite is executed in GH Actions.
Should we accept these as baseline or set `GITHUB_ACTIONS` to `false` in CI tests?
Accepting them as a baseline will likely make it inconvenient to run the tests locally.

ematipico commented:
> @ematipico the tests are failing due to the workflow commands being emitted because the whole test suite is executed in GH Actions. Should we accept these as baseline or set `GITHUB_ACTIONS` to `false` in CI tests? Accepting them as a baseline will likely make it inconvenient to run the tests locally.

We should disable this functionality in the snapshot tests; as you said, it would create a lot of friction. Although it would be great if we could have some assertions before printing the snapshot, to make sure that the output is correct.

We can use the `redact_snapshot` function inside `snap_test.rs` to remove that message from the final snapshot.

nikeee commented:
@ematipico I've added it to the redaction in a9d5572d785ccd95d03cf6b07b1e1b139a60e825. There are cases where a workflow command is emitted without a "normal" diagnostic message, which results in the output being empty. To match against current snapshots, I've refactored the method to return an `Option<T>` and skipped the message as if it wasn't there to begin with.
